azerothcore/azerothcore-wotlk
# feat(Core/Unit): New helper HasActivePowerType and script hook OnPlayerHasActivePowerType

## Changes Proposed:
This PR proposes changes to:

Added a new hook "OnPlayerHasActivePowerType" inside a new interception helper function "Player::HasActivePowerType" (which has a virtual base Unit::HasActivePowerType) and replaced many comparison uses of Unit::getPowerType() == power.  This hook allows active power type overrides granting the ability to use any class power you choose on any other class, such as rage on a Mage or mana on a Warrior.

## Issues Addressed:
none

## SOURCE:
n/a

## Tests Performed:
This PR has been:
- [X] Tested in-game by the author (and used on a personal server)

## How to Test the Changes:
To test the change impact to core/stock: use rage/energy/mana/runic on the respective classes.  

To test this change as a hooked script, implement the PlayerScript hook OnPlayerHasActivePowerType.  An example of this implementation can be found in this mod branch I'm working on here: https://github.com/NathanHandley/mod-multi-class/blob/NewClassChecks/src/MultiClass_PlayerScript.cpp (line 78).  From there, give yourself abilities like Heroic Strike on a Paladin to see the effect.

## Known Issues and TODO List:
- This PR does not introduce any known issues, but many script-overridden cross-class abilities (notably Death Knight rune powers and Hunter/Warrior/Rogue reactives) won't function without also merging in this PR: https://github.com/azerothcore/azerothcore-wotlk/pull/18243.
